clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -124,8 +182,72 @@ public KafkaBasedCommandSender(
         Objects.requireNonNull(deviceId);
         Objects.requireNonNull(command);
 
-        // TODO to implement.
-        return null;
+        final Span span = TracingHelper
+                .buildChildSpan(tracer, context, "send command and wait for response", getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .withTag(TracingHelper.TAG_TENANT_ID, tenantId)
+                .withTag(TracingHelper.TAG_DEVICE_ID, deviceId)
+                .start();
+        final String correlationId = correlationIdSupplier.get();
+        final Promise<Void> requestTracker = Promise.promise();
+        final Promise<DownstreamMessage<KafkaMessageContext>> responseTracker = Promise.promise();
+        final Promise<Void> timeOutTracker = Promise.promise();
+        final long timerId = vertx.setTimer(COMMAND_TIMEOUT, id -> {
+            timeOutTracker.fail(new SendMessageTimeoutException("sending command/waiting for response timed out after "
+                    + COMMAND_TIMEOUT + "ms"));
+            requestTracker.tryComplete();
+            responseTracker.tryComplete();
+        });
+        final Handler<DownstreamMessage<KafkaMessageContext>> messageHandler = m -> {
+            //if the correlation id of the received response matches with that of the command
+            if (m.getCorrelationId() != null) {
+                Optional.ofNullable(commandResponseSubscriptions.get(tenantId))
+                        .map(Pair::two)
+                        .filter(ids -> ids.remove(m.getCorrelationId()))
+                        .ifPresent(ok -> {
+                            vertx.cancelTimer(timerId);
+                            span.log("received command response");
+                            LOGGER.debug("received command response and cancelled timer.");
+                            responseTracker.tryComplete(m);
+                            timeOutTracker.tryComplete();
+                        });
+            }
+        };
+
+        subscribeForCommandResponse(tenantId, messageHandler, span)
+                .compose(ok -> sendCommand(tenantId, deviceId, command, contentType, data, correlationId, properties,
+                        true, span.context())
+                .onSuccess(sent -> {
+                    commandResponseSubscriptions.get(tenantId).two().add(correlationId);
+                    span.log("sent command and waiting for response");
+                    LOGGER.debug("sent command [correlation-id: {}] and waiting for response", correlationId);
+                    requestTracker.complete();
+                })
+                .onFailure(error -> {
+                    TracingHelper.logError(span, "error sending command", error);
+                    LOGGER.debug("error sending command", error);
+                    requestTracker.fail(error);
+                    responseTracker.tryComplete();
+                    timeOutTracker.tryComplete();
```

Comment:
`CompositeFuture.all()` will fail and when at-least one future fails. But it doesn't mean the other 2 futures are also get completed. The other futures still await for completion and I thought of proactively completing them.

Reply:
> But it doesn't mean the other 2 futures are also get completed. The other futures still await for completion and I thought of proactively completing them.

But there's nothing waiting on these futures after `CompositeFuture.all()` has failed, so whether these futures get completed or not doesn't matter.

Reply:
> so whether these futures get completed or not doesn't matter.

Since we already knew that these futures need not wait any further and FMPOV it's good to proactively complete them. I don't see any disadvantages in completing them.